// mjml-accordion/lib/AccordionText.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _mjmlCore = require('mjml-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MjAccordionText = (_temp = _class = function (_BodyComponent) {
  (0, _inherits3.default)(MjAccordionText, _BodyComponent);

  function MjAccordionText() {
    (0, _classCallCheck3.default)(this, MjAccordionText);
    return (0, _possibleConstructorReturn3.default)(this, (MjAccordionText.__proto__ || (0, _getPrototypeOf2.default)(MjAccordionText)).apply(this, arguments));
  }

  (0, _createClass3.default)(MjAccordionText, [{
    key: 'getStyles',
    value: function getStyles() {
      return {
        td: {
          background: this.getAttribute('background-color'),
          'font-size': this.getAttribute('font-size'),
          'font-family': this.getAttribute('font-family'),
          color: this.getAttribute('color'),
          'padding-bottom': this.getAttribute('padding-bottom'),
          'padding-left': this.getAttribute('padding-left'),
          'padding-right': this.getAttribute('padding-right'),
          'padding-top': this.getAttribute('padding-top'),
          padding: this.getAttribute('padding')
        },
        table: {
          width: '100%',
          'border-bottom': this.getAttribute('border')
        }
      };
    }
  }, {
    key: 'renderContent',
    value: function renderContent() {
      return '\n      <td\n        ' + this.htmlAttributes({
        class: this.getAttribute('css-class'),
        style: 'td'
      }) + '\n      >\n        ' + this.getContent() + '\n      </td>\n    ';
    }
  }, {
    key: 'render',
    value: function render() {
      return '\n      <div\n        ' + this.htmlAttributes({
        class: 'mj-accordion-content'
      }) + '\n      >\n        <table\n          ' + this.htmlAttributes({
        'cell-spacing': '0',
        'cell-padding': '0',
        style: 'table'
      }) + '\n        >\n          <tbody>\n            <tr>\n              ' + this.renderContent() + '\n            </tr>\n          </tbody>\n        </table>\n      </div>\n    ';
    }
  }]);
  return MjAccordionText;
}(_mjmlCore.BodyComponent), _class.endingTag = true, _class.allowedAttributes = {
  'background-color': 'color',
  'font-size': 'unit(px,%)',
  'font-family': 'string',
  color: 'color',
  'padding-bottom': 'unit(px,%)',
  'padding-left': 'unit(px,%)',
  'padding-right': 'unit(px,%)',
  'padding-top': 'unit(px,%)',
  padding: 'unit(px,%){1,4}'
}, _class.defaultAttributes = {
  'font-size': '13px',
  padding: '16px'
}, _temp);
exports.default = MjAccordionText;
module.exports = exports['default'];
