// mjml-body/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _mjmlCore = require('mjml-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MjBody = (_temp = _class = function (_BodyComponent) {
  (0, _inherits3.default)(MjBody, _BodyComponent);

  function MjBody() {
    (0, _classCallCheck3.default)(this, MjBody);
    return (0, _possibleConstructorReturn3.default)(this, (MjBody.__proto__ || (0, _getPrototypeOf2.default)(MjBody)).apply(this, arguments));
  }

  (0, _createClass3.default)(MjBody, [{
    key: 'getChildContext',
    value: function getChildContext() {
      return (0, _extends3.default)({}, this.context, {
        containerWidth: this.getAttribute('width')
      });
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      return {
        div: {
          'background-color': this.getAttribute('background-color')
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var setBackgroundColor = this.context.setBackgroundColor;

      setBackgroundColor(this.getAttribute('background-color'));

      return '\n      <div\n        ' + this.htmlAttributes({
        class: this.getAttribute('css-class'),
        style: 'div'
      }) + '\n      >\n        ' + this.renderChildren() + '\n      </div>\n    ';
    }
  }]);
  return MjBody;
}(_mjmlCore.BodyComponent), _class.defaultAttributes = {
  width: '600px'
}, _temp);
exports.default = MjBody;
module.exports = exports['default'];
