// mjml-wrapper/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _mjmlSection = require('mjml-section');

var _mjmlSection2 = _interopRequireDefault(_mjmlSection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MjWrapper = function (_MjSection) {
  (0, _inherits3.default)(MjWrapper, _MjSection);

  function MjWrapper() {
    (0, _classCallCheck3.default)(this, MjWrapper);
    return (0, _possibleConstructorReturn3.default)(this, (MjWrapper.__proto__ || (0, _getPrototypeOf2.default)(MjWrapper)).apply(this, arguments));
  }

  (0, _createClass3.default)(MjWrapper, [{
    key: 'renderWrappedChildren',
    value: function renderWrappedChildren() {
      var children = this.props.children;
      var containerWidth = this.context.containerWidth;


      return '\n      ' + this.renderChildren(children, {
        renderer: function renderer(component) {
          return component.constructor.isRawElement() ? component.render() : '\n          <!--[if mso | IE]>\n            <tr>\n              <td\n                ' + component.htmlAttributes({
            align: component.getAttribute('align'),
            class: component.getAttribute('css-class') ? component.getAttribute('css-class').split(' ').map(function (c) {
              return c + '-outlook';
            }).join(' ') : null,
            width: containerWidth
          }) + '\n              >\n          <![endif]-->\n            ' + component.render() + '\n          <!--[if mso | IE]>\n              </td>\n            </tr>\n          <![endif]-->\n        ';
        }
      }) + '\n    ';
    }
  }]);
  return MjWrapper;
}(_mjmlSection2.default);

exports.default = MjWrapper;
module.exports = exports['default'];
