// mjml-spacer/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp;

var _mjmlCore = require('mjml-core');

var _conditionalTag = require('mjml-core/lib/helpers/conditionalTag');

var _conditionalTag2 = _interopRequireDefault(_conditionalTag);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MjSpacer = (_temp = _class = function (_BodyComponent) {
  (0, _inherits3.default)(MjSpacer, _BodyComponent);

  function MjSpacer() {
    (0, _classCallCheck3.default)(this, MjSpacer);
    return (0, _possibleConstructorReturn3.default)(this, (MjSpacer.__proto__ || (0, _getPrototypeOf2.default)(MjSpacer)).apply(this, arguments));
  }

  (0, _createClass3.default)(MjSpacer, [{
    key: 'getStyles',
    value: function getStyles() {
      return {
        div: {
          height: this.getAttribute('height')
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      var height = this.getAttribute('height');

      return '\n      ' + (0, _conditionalTag2.default)('\n        <table role="presentation" border="0" cellpadding="0" cellspacing="0"><tr><td height="' + parseInt(height, 10) + '" style="vertical-align:top;height:' + height + ';">\n      ') + '\n      <div\n        ' + this.htmlAttributes({
        style: 'div'
      }) + '\n      >\n        &nbsp;\n      </div>\n      ' + (0, _conditionalTag2.default)('\n        </td></tr></table>\n      ') + '\n    ';
    }
  }]);
  return MjSpacer;
}(_mjmlCore.BodyComponent), _class.allowedAttributes = {
  border: 'unit(px)',
  'border-bottom': 'unit(px)',
  'border-left': 'unit(px)',
  'border-radius': 'unit(px)',
  'border-right': 'unit(px)',
  'border-top': 'unit(px)',
  'container-background-color': 'color',
  direction: 'enum(ltr,rtl)',
  'padding-bottom': 'unit(px,%)',
  'padding-left': 'unit(px,%)',
  'padding-right': 'unit(px,%)',
  'padding-top': 'unit(px,%)',
  padding: 'unit(px,%){1,4}',
  'vertical-align': 'string',
  width: 'unit(px,%)',
  height: 'unit(px,%)'
}, _class.defaultAttributes = {}, _temp);
exports.default = MjSpacer;
module.exports = exports['default'];
